### PR TITLE
Support gather of pytorch

### PR DESCRIPTION
**Pull request by KexinFeng**
## Description ##

This originiates from issue https://github.com/deepjavalibrary/djl/issues/248. This is regarding support of advanced indexing.

As part of this support, this PR implements `gather`. It's the same as the gather in
https://pytorch.org/docs/1.11/generated/torch.gather.html#torch.gather 
https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html
It currently supports pytorch engine only. MXNet and TensorFlow have different defination of gather.

The design is the following:
`gather(arr, idx, axis)` returns the following based on the conventions [1](https://pytorch.org/docs/1.11/generated/torch.gather.html#torch.gather) [2](https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html
):
```
out_{ijk} = arr_{idx_{ijk}, j, k} if axis=0
              or arr_{i, idx_{ijk}, k} if axis=1
              or arr_{i, j, idx_{ijk}}  if axis=2
```
But `gather` is used to index across axis, including the preceding axis. This *conflicts* with the linear storage of `List<NDIndexElement>` inside `NDIndex`. So `gather` is treated parallel to `NDIndex` when fed into `get`.

Similar functions in numpy / pytorch / mxnet can also be implemented: e.g. `torch.take`, and advanced indexing like `arr[:, [0 ,2]]` and `arr[idx_r, idx_c]`

### CODE REVIEW

File: api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java:2]:
```suggestion
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import ai.djl.ndarray.NDArray;
+
+/** A simplified representation of a gather-based {@link NDIndex}-like class. */
+public final class NDIndexFullGather {
+
+    private NDArray indices;
+    private int axis;
+
+    /**
+     * Constructs a new {@link NDIndexFullGather}.
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java:24]:
```suggestion
     * Constructs a new {@code NDIndexFullGather} instance.
```

Comment (KexinFeng) [reply api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java]: Got it. Thanks!

File: api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java (hunk 1 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/ndarray/index/NDIndexFullGather.java]: Got it. Thanks!

File: integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java [hunk 3]
```diff
@@ -52,6 +54,23 @@ public void testPick() {
         }
     }
 
+    @Test
+    public void testGather() {
+        // Currently in windows gradle cannot find all the engines.
+        // In the dependency, changing runtimeOnly to api however will remedy the problem.
+        // TODO: remove this when gradle problem is fixed.
+        TestRequirements.notWindows();
+        Engine engine = Engine.getEngine("PyTorch");
```
Comment (zachgk) [inline integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java:63]: The integration module is supposed to be for "engine agnostic tests". So, you shouldn't need to specify "PyTorch" here. For the other engines, it will throw a `UnsupportedOperationException` instead of running successfully. But, our test runner is designed to treat `UnsupportedOperationException` as a special test result, not a failure. It's similar to how a testng `SkipException` counts as a skipped test result rather than a test failure.

In the future, we should be able to implement gather for the other engines and then it will just run the test because it is already there.

**Review by zachgk — APPROVED** [review]

File: integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java [hunk 4]
```diff
@@ -52,6 +53,22 @@ public void testPick() {
         }
     }
 
+    @Test
+    public void testGather() {
+        // Currently in windows gradle cannot find all the engines.
```
Comment (frankfliu) [inline integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java:58]: Is this comment still valid?

Comment (zachgk) [reply integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java:58]: Probably. It seems like something is wrong with `runtimeOnly` dependencies on windows gradle.

Comment (KexinFeng) [reply integration/src/main/java/ai/djl/integration/tests/ndarray/NDIndexTest.java:58]: Yes after updating gradle, on GitHub the test stills fails for the same reason
